src/rcc.rs

```diff
@@ -214,92 +232,112 @@ impl CFGR {
     }
 
     /// Returns a tuple containing the effective sysclk rate and optional pll settings.
-    fn calc_sysclk(&self) -> (u32, Option<(u8, rcc::cfgr::PLLSRC_A)>) {
+    fn calc_sysclk(&self) -> (u32, Option<(cfgr::PLLMUL_A, cfgr::PLLSRC_A)>) {
         let (pllsrcclk, pllmul, pllsrc) = self.calc_pll();
         if pllmul == 1 {
             return (pllsrcclk, None);
         }
 
         let pllmul = cmp::min(cmp::max(pllmul, 2), 16);
-        let sysclk = pllmul * pllsrcclk;
+        let sysclk = pllsrcclk * pllmul;
         assert!(sysclk <= 72_000_000);
 
-        let pllmul_bits = pllmul as u8 - 2;
-        (sysclk, Some((pllmul_bits, pllsrc)))
+        // NOTE From<u8> for PLLMUL_A is not implemented
+        let pllmul_mul = match pllmul as u8 {
+            2 => cfgr::PLLMUL_A::MUL2,
+            3 => cfgr::PLLMUL_A::MUL3,
+            4 => cfgr::PLLMUL_A::MUL4,
+            5 => cfgr::PLLMUL_A::MUL5,
+            6 => cfgr::PLLMUL_A::MUL6,
+            7 => cfgr::PLLMUL_A::MUL7,
+            8 => cfgr::PLLMUL_A::MUL8,
+            9 => cfgr::PLLMUL_A::MUL9,
+            10 => cfgr::PLLMUL_A::MUL10,
+            11 => cfgr::PLLMUL_A::MUL11,
+            12 => cfgr::PLLMUL_A::MUL12,
+            13 => cfgr::PLLMUL_A::MUL13,
+            14 => cfgr::PLLMUL_A::MUL14,
+            15 => cfgr::PLLMUL_A::MUL15,
+            16 => cfgr::PLLMUL_A::MUL16,
+            _ => unreachable!(),
+        };
+        (sysclk, Some((pllmul_mul, pllsrc)))
     }
 
     /// Freezes the clock configuration, making it effective
     pub fn freeze(self, acr: &mut ACR) -> Clocks {
         let (sysclk, pll_options) = self.calc_sysclk();
 
-        let hpre_bits = self
+        // TODO try this method
+        // https://github.com/stm32-rs/stm32f4xx-hal/blob/c888eaa4d7dcac9670102896e54945efb5c41a3b/src/rcc.rs#L263
+        let (hpre_bits, hpre_div) = self
             .hclk
             .map(|hclk| match sysclk / hclk {
                 0 => unreachable!(),
-                1 => 0b0111,
-                2 => 0b1000,
-                3..=5 => 0b1001,
-                6..=11 => 0b1010,
-                12..=39 => 0b1011,
-                40..=95 => 0b1100,
-                96..=191 => 0b1101,
-                192..=383 => 0b1110,
-                _ => 0b1111,
+                1 => (cfgr::HPRE_A::DIV1, 1),
+                2 => (cfgr::HPRE_A::DIV2, 2),
+                3..=5 => (cfgr::HPRE_A::DIV4, 4),
+                6..=11 => (cfgr::HPRE_A::DIV8, 8),
+                12..=39 => (cfgr::HPRE_A::DIV16, 16),
+                40..=95 => (cfgr::HPRE_A::DIV64, 64),
+                96..=191 => (cfgr::HPRE_A::DIV128, 128),
+                192..=383 => (cfgr::HPRE_A::DIV256, 256),
+                _ => (cfgr::HPRE_A::DIV512, 512),
             })
-            .unwrap_or(0b0111);
+            .unwrap_or((cfgr::HPRE_A::DIV1, 1));
 
-        let hclk = sysclk / (1 << (hpre_bits - 0b0111));
+        let hclk: u32 = sysclk / 1 << hpre_div;
```

Comment:
I don't understand this line.  I was expecting `let hclk = sysclk / hpre_div;`

Reply:
You are right. The bitshift is not needed anymore, as the values, which where previously calculated from the bitshift, are explicitly listed now. I'll fix this.